Refactoring in coverage instrumentation

romainbrenguier opened the pull request:
This contains a few commit whose intent are to improve clarity of the code in the coverage instrumentation.

romainbrenguier committed:
Make format_number_range function instead of class

romainbrenguier committed:
Make representative_inst an optional

This is more explicit than using goto_program.instructions.end to denote
the empty case.

romainbrenguier committed:
Make update_covered_lines a static function

This is only working on the block_info member.

romainbrenguier committed:
Declare constants as const

romainbrenguier committed:
Pull continuation_of_block function out

We pull this function out of the cover_basic_blockst constructor, as the
constructor is rather long, and having a function for this part of the
code makes it clearer.

romainbrenguier committed:
Define constants as const

LAJW commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -19,6 +19,22 @@ Author: Peter Schrammel
+  if(instruction->incoming_edges.size() != 1)
+    return {};
+
+  goto_programt::targett in_t = *instruction->incoming_edges.cbegin();
```
Could be `const`.

chrisr-diffblue commented on src/goto-instrument/cover_basic_blocks.h:
```diff
@@ -21,6 +21,20 @@ class message_handlert;
+    optionalt<goto_programt::const_targett> representative_inst;
+
+    /// the source location representative for this block
+    // (we need a separate copy of source locations because we attach
```
Tiny nit: Mix of '///' and '//' comments here

chrisr-diffblue reviewed (approved):
Looks OK to me, with a tiny formatting nit.

smowton commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -15,6 +15,26 @@ Author: Peter Schrammel
 
+/// create list of covered lines as CSV string and set as property of source
+/// location of basic block, compress to ranges if applicable
+static void update_covered_lines(cover_basic_blockst::block_infot &block_info);
```
I think you could just define this here rather than forward-declare it?

romainbrenguier replied:
Yes it was mostly to avoid moving too much code, have a smaller diff and make it easier to review

LAJW replied:
Which is a Good Thing, anyway. It's much easier to just glance at the definitions at the top of the file than to scroll through the whole thing trying to create a mental map of its contents. Forward declarations are there to avoid function repositioning every time we want to use it before it's declared. I'd leave it as it is.

smowton reviewed (approved):
Looks good, one nitpick

peterschrammel commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -214,7 +218,7 @@ void cover_basic_blockst::output(std::ostream &out) const
 }
 
-void cover_basic_blockst::update_covered_lines(block_infot &block_info)
+static void update_covered_lines(cover_basic_blockst::block_infot &block_info)
```
But now block_infot gets unnecessarily exposed.

peterschrammel commented on src/goto-instrument/cover_basic_blocks.h:
```diff
@@ -62,27 +77,9 @@ class cover_basic_blockst
 
-  /// create list of covered lines as CSV string and set as property of source
-  /// location of basic block, compress to ranges if applicable
-  void update_covered_lines(block_infot &block_info);
```
You could make this static inside the class instead of exposing block_infot.

peterschrammel commented on src/goto-instrument/cover_basic_blocks.cpp:
```diff
@@ -19,6 +19,22 @@ Author: Peter Schrammel
+/// forward gotos, return this blocks number.
+static optionalt<unsigned> continuation_of_block(
+  const goto_programt::const_targett &instruction,
+  std::map<goto_programt::const_targett, unsigned> &block_map)
```
If this method was inside the class you could take advantage of the block_map typedef (it would be used twice then).

peterschrammel commented on src/goto-instrument/cover_basic_blocks.h:
```diff
@@ -23,15 +23,14 @@ class cover_basic_blockst
   struct block_infot
   {
-    /// the program location to instrument for this block
+    /// the program instruction representative for this block
```
There a couple of typos in the commit message.

peterschrammel commented on src/goto-instrument/cover_basic_blocks.h:
```diff
@@ -20,6 +20,33 @@ class message_handlert;
 
 class cover_basic_blockst
 {
+private:
```
Merge the private sections at the end of the class.

romainbrenguier replied:
@peterschrammel Is there a reason for that? I'd prefer to have type definitions at the beginning of the class.

peterschrammel replied:
My view on this is that I'd like to see the interface of the class first. Internal stuff can come at the end...
Of course, if there are type definitions that are part of the public interface then they should go first.

romainbrenguier committed:
Rearrange cover_basic_blocks header

Use private instead of protected.
Remove typedef used only once.
Document members.
Mark class final.
